(docs): add blog post on snapshot testing (#12537)
webdriverio/webdriverio: website/blog/2024-03-20-snapshot-testing.md

/>

Let's dive into each of these powerful capabilities.

## DOM or Object Snapshots

For evaluating the state of the DOM, a large object or the content of a UI element we often tend to copy the value into our test and manually update it if we change the behavior of our application or component.

With text-based snapshots, we can just have this handled by WebdriverIO. For example, let's say we want to verify the state of our React component in a browser, we can just do the following:

```tsx title="/src/component.test.tsx"
import { expect, $ } from '@wdio/globals'
import { render } from '@testing-library/react'

function App() {
    const [theme, setTheme] = useState('light')

    const toggleTheme = () => {
        const nextTheme = theme === 'light' ? 'dark' : 'light'
        setTheme(nextTheme)
    }

    return <button onClick={toggleTheme}>
        Current theme: {theme}
    </button>
}

describe('React Component Testing', () => {
    it('supports snapshot tests', async () => {
        const { container } = render(<App />)
        await expect(container).toMatchSnapshot()
        await $('button').click()
        await expect(container).toMatchSnapshot()
    })
})
```

> WebdriverIO will automatically grab the DOM structure of the component and store a snapshot file in `/src/__snapshots__` directory next to your test.
> 
> component.test.tsx.snap` with the following content:

Reviewer: ```suggestion
WebdriverIO will automatically grab the DOM structure of the component and store a snapshot file called `component.test.tsx.snap` in `/src/__snapshots__` directory next to your test with the following content:
```

exports[`React Component Testing > supports snapshot tests 2`] = `"<div><button>Current theme: dark</button></div>"`;
```

If you prefer to keep the snapshots as part of your tests you can use `toMatchInlineSnapshot` instead:

```tsx
await expect(container).toMatchInlineSnapshot()
```

After running the test for the first time, WebdriverIO will make a change to the test and fill in the snapshot inline:

```tsx
await expect(container).toMatchInlineSnapshot(`"<div><button>Current theme: light</button></div>"`)
```

Now, if you make changes to your component that will impact all snapshots you can update them all in a single run by calling:

```sh
npx wdio run wdio.conf.ts --updateSnapshots
# or
npx wdio run wdio.conf.ts -s
```

This makes maintaining your tests so much easier. The same works for all other types of objects, e.g. CSS Properties, or the text content of an element. They all can be converted into a snapshot to simplify the assertion and keep your tests lean. This can also speed up your tests by merging many single assertions into one:

```ts
const elem = $('#alertBar')
await expect(elem).toHaveAttribute('data-alert')

> await expect(elem).toHaveCSSProperty('background-color', 'green')

Reviewer: This is not covered by the snapshot, we don't get a complete DOM snapshot with computed styles

await expect(elem).toHaveText('You logged into a secure area!')
```

now becomes a single:

```ts
await expect($('#alertBar')).toMatchSnapshot()
/**
 * stores the following into a snapshot file:
 *
 * <div data-alert="" id="flash" class="flash success">
 *   You logged into a secure area!
 *   <a href="#" class="close">×</a>
 * </div>
 */
```

You can find more information about DOM and object-based snapshots in our [Snapshot guide](/docs/snapshot).

## Visual Snapshots

> While taking snapshots of an element structure and its attributes might be great and powerful, it comes with an important caveat: even though we are testing that the element has a class name called `success`, this doesn't guarantee that the alert is green!

Reviewer: This confirms my above comment

- take a visual snapshot of the whole screen:
  ```ts
  await expect(browser).toMatchScreenSnapshot('partialPage')
  ```
- take a visual snapshot of an element:
  ```ts
  await expect($('#element-id')).toMatchElementSnapshot('firstButtonElement')
  ```
- take a snapshot of the whole page:
  ```ts
  await expect(browser).toMatchFullPageSnapshot('fullPage')
  ```
- or take a page snapshot that includes [page tab-ability](https://vivrichards.co.uk/accessibility/automating-page-tab-flows-using-visual-testing-and-javascript):
  ```ts
  await expect(browser).toMatchTabbablePageSnapshot('check-tabbable')
  ```

WebdriverIO will store these visual snapshots conveniently next to your text-based snapshots within the `__snapshots__` directory next to your test.

While text-based snapshot testing is built into WebdriverIO, you have to install a [service](https://www.npmjs.com/package/@wdio/visual-service) to enable all visual snapshot capabilities via:

```sh
npm i --save-dev @wdio/visual-service
```

> With the most recent release of the Visual Testing Module with have shipped further improvements for **Native App Snapshot Testing**.

Reviewer: ```suggestion
With the most recent release of the Visual Testing Module with have shipped further improvements for **Mobile Native App Snapshot Testing**.
```

> ### Native App Snapshot Testing

Reviewer: ```suggestion
### Mobile Native App Snapshot Testing
```

> The module now supports `saveElement`, `checkElement`, `saveScreen`, and `checkScreen` methods, along with `toMatchElementSnapshot` and `toMatchScreenSnapshot` matchers for native apps. It automatically detects the testing context (web, webview, or native_app) to streamline your workflow.

Reviewer: The blog never talked about the "internal" methods, so maybe adjust it to the following

```suggestion
The module now supports the `toMatchElementSnapshot` and `toMatchScreenSnapshot` matchers for Mobile native apps. It automatically detects the testing context (web, webview, or native_app) to streamline your workflow.
```

### Key Features of the Visual Service

Some of the features that make visual testing with WebdriverIO unique are:

- save or compare __screens/elements/full-page__ screens against a baseline
- automatically __create a baseline__ when no baseline is there
- __block out custom regions__ and even __automatically exclude__ a status and or toolbars (mobile only) during a comparison
- increase the element dimensions screenshots
- __hide text__ during website comparison to:
  - __improve stability__ and prevent font rendering flakiness
  - only focus on the __layout__ of a website
- use __different comparison methods__ and a set of __additional matchers__ for better readable tests
- verify how your website will __support tabbing with your keyboard__, see also [Tabbing through a website](visual-testing#tabbing-through-a-website)
- and much more, see the [service](/docs/visual-testing/service-options) and [method](/docs/visual-testing/method-options) options

> Learn everything about WebdriverIO visual testing capabilities in our [Visual docs](visual-testing).

Reviewer: Maybe also link to the Visual Testing discord channel

### Special Thanks to [`@wswebcreation`](https://github.com/wswebcreation)

> We owe a big thank you to our core maintainer [Wim Selles](https://github.com/wswebcreation) for his work in [`wdio-native-app-compare`](https://github.com/wswebcreation/wdio-native-app-compare), which inspired this enhancement. His contribution has been vital in advancing our module's capabilities.

Reviewer: ❤️
